core/common/src/main/java/alluxio/network/protocol/databuffer/RefCountedNioByteBuf.java

```diff
@@ -0,0 +1,409 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.network.protocol.databuffer;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.AbstractReferenceCountedByteBuf;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.Unpooled;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.channels.ScatteringByteChannel;
+
+/**
+ * Reference counted NIO byte buffer wrapped in Netty ByteBuf.
+ */
+abstract class RefCountedNioByteBuf extends AbstractReferenceCountedByteBuf {
+  protected final ByteBuffer mDelegate;
+  protected int mCapacity;
+
+  protected RefCountedNioByteBuf(ByteBuffer buffer, int capacity) {
+    super(capacity);
+    Preconditions.checkArgument(buffer.capacity() >= capacity);
+    buffer.clear();
+    // enforce big endianness
+    buffer.order(ByteOrder.BIG_ENDIAN);
+    mDelegate = buffer;
+    mCapacity = capacity;
+  }
+
+  @Override
+  protected byte _getByte(int index) {
+    return mDelegate.get(index);
+  }
+
+  @Override
+  protected short _getShort(int index) {
+    return mDelegate.getShort(index);
+  }
+
+  @Override
+  protected short _getShortLE(int index) {
+    return Short.reverseBytes(_getShort(index));
+  }
+
+  @Override
+  protected int _getUnsignedMedium(int index) {
+    byte first = _getByte(index);
+    byte second = _getByte(index + 1);
+    byte third = _getByte(index + 2);
+    return (Byte.toUnsignedInt(first) << 16)
+        | (Byte.toUnsignedInt(second) << 8)
+        | Byte.toUnsignedInt(third);
+  }
+
+  @Override
+  protected int _getUnsignedMediumLE(int index) {
```

Comment:
do we really need all of these `_` prefix methods?

Reply:
these are abstract methods defined by `AbstractByteBuf`. I didn't find an appropriate base class that has already implemented them.